### PR TITLE
[Mellanox] Extend mellanox platform API to report SFP error event

### PR DESCRIPTION
**Pull request by keboliu**
**- What I did**
Extend SFP event API to report SFP error to the upper layer app.

**- How I did it**
handle SFP event which reported from SDK, log all the detail error type and pass the error event that will block EEPROM accessing to the upper layer.

**- How to verify it**
test on Mellanox platform

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

**This PR has a dependency on https://github.com/Azure/sonic-platform-common/pull/80**

### CODE REVIEW

**Comment by qiluo-msft** [comment]
>                 if not port_dict == {}:

If not bool(port_dict):

---
Refers to: platform/mellanox/mlnx-platform-api/sonic_platform/chassis.py:451 in 3cef25c. [](commit_id = 3cef25ca3d0e9c874d70d88d0fce2c694535f29a, deletion_comment = False)

File: platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py [hunk 1]
```diff
@@ -11,15 +11,69 @@
 from python_sdk_api.sx_api import *
 from sonic_daemon_base.daemon_base import Logger
 
-SDK_SFP_STATE_IN = 0x1
+# SFP status from PMAOS register
+# 0x1 plug in
+# 0x2 plug out
+# 0x3 plug in with error
+# 0x4 disabled, at this status SFP eeprom is not accessable,
```
Comment (qiluo-msft) [inline platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py:18]: >accessable [](start = 49, length = 10)

accessible

File: platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py [hunk 2]
```diff
@@ -11,15 +11,69 @@
...
+
+# SFP status that will be handled by XCVRD 
 STATUS_PLUGIN = '1'
 STATUS_PLUGOUT = '0'
-STATUS_UNKNOWN = '2'
+STATUS_ERR_I2C_STUCK = '2'
+STATUS_ERR_BAD_EEPROM = '3'
+STATUS_ERR_UNSUPPORTED_CABLE = '4'
+STATUS_ERR_HIGH_TEMP = '5'
+STATUS_ERR_BAD_CABLE = '6'
```
Comment (qiluo-msft) [inline platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py:33]: The list is common to all vendors, why repeat for Mellanox?

File: platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py [hunk 3]
```diff
@@ -11,15 +11,69 @@
...
+    0x5: STATUS_ERR_UNSUPPORTED_CABLE,
+    0x6: STATUS_ERR_HIGH_TEMP,
+    0x7: STATUS_ERR_BAD_CABLE
+}
 
 sfp_value_status_dict = {
         SDK_SFP_STATE_IN:  STATUS_PLUGIN,
         SDK_SFP_STATE_OUT: STATUS_PLUGOUT,
+        SDK_SFP_STATE_ERR: STATUS_ERROR,
+        SDK_SFP_STATE_DIS: STATUS_PLUGOUT,
```
Comment (qiluo-msft) [inline platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py:76]: less indentations?

File: platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py [hunk 4]
```diff
@@ -174,12 +228,13 @@ def check_sfp_status(self, port_change, timeout):
...
         for fd in read:
             if fd == self.rx_fd_p.fd:
-                success, port_list, module_state = self.on_pmpe(self.rx_fd_p)
+                success, port_list, module_state, error_type = self.on_pmpe(self.rx_fd_p)
                 if not success:
                     logger.log_error("failed to read from {}".format(fd))
                     break
 
                 sfp_state = sfp_value_status_dict.get(module_state, STATUS_UNKNOWN)
+                sfp_event = sfp_state
```
Comment (qiluo-msft) [inline platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py:237]: >sfp_event = sfp_state [](start = 16, length = 21)

This is hard to read, could you remove?

**Comment by qiluo-msft** [comment]
>             return True

return found != 0;

---
Refers to: platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py:268 in 3cef25c. [](commit_id = 3cef25ca3d0e9c874d70d88d0fce2c694535f29a, deletion_comment = False)

File: platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py [hunk 5]
```diff
@@ -228,7 +293,17 @@ def on_pmpe(self, fd_p):
             port_list_size = pmpe_t.list_size
             logical_port_list = pmpe_t.log_port_list
             module_state = pmpe_t.module_state
-
+            error_type = pmpe_t.error_type
+            module_id = pmpe_t.module_id
+
+            if module_state == SDK_SFP_STATE_ERR:
+                logger.log_error("Receive PMPE error event on module {}: status {} error type {} ".format(module_id, module_state, error_type))
```
[the remark this answers is not in the document]
Comment (keboliu) [reply platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py]: fixed

File: platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py [hunk 6]
```diff
@@ -228,7 +293,17 @@ def on_pmpe(self, fd_p):
...
+            elif module_state == SDK_SFP_STATE_DIS:
+                logger.log_info("Receive PMPE disable event on module {}: status {} ".format(module_id, module_state))
+            elif module_state == SDK_SFP_STATE_IN or module_state == SDK_SFP_STATE_OUT:
+                logger.log_info("Receive PMPE plug in/out event on module {}: status {} ".format(module_id, module_state))
```
[the remark this answers is not in the document]
Comment (keboliu) [reply platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py]: fixed

File: platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py [hunk 7]
```diff
@@ -228,7 +293,17 @@ def on_pmpe(self, fd_p):
...
             logical_port_list = pmpe_t.log_port_list
             module_state = pmpe_t.module_state
-
+            error_type = pmpe_t.error_type
+            module_id = pmpe_t.module_id
+
+            if module_state == SDK_SFP_STATE_ERR:
+                logger.log_error("Receive PMPE error event on module {}: status {} error type {} ".format(module_id, module_state, error_type))
+            elif module_state == SDK_SFP_STATE_DIS:
+                logger.log_info("Receive PMPE disable event on module {}: status {} ".format(module_id, module_state))
```
[the remark this answers is not in the document]
Comment (keboliu) [reply platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py]: fixed

File: platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py [hunk 8]
```diff
@@ -228,7 +293,17 @@ def on_pmpe(self, fd_p):
...
+            elif module_state == SDK_SFP_STATE_IN or module_state == SDK_SFP_STATE_OUT:
+                logger.log_info("Receive PMPE plug in/out event on module {}: status {} ".format(module_id, module_state))
+            else:
+                logger.log_error("Receive PMPE unknown event on module {}: status {} ".format(module_id, module_state))
```
[the remark this answers is not in the document]
Comment (keboliu) [reply platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py]: fixed

File: platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py (hunk 1 above)
Comment (keboliu) [reply platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py]: fixed

File: platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py (hunk 3 above)
Comment (keboliu) [reply platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py]: fixed

File: platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py (hunk 4 above)
Comment (keboliu) [reply platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py]: fixed

File: platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py (hunk 2 above)
Comment (keboliu) [reply platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py:33]: On Mellanox platform the error code values is not exactly the same as commonly defined error code values, need to map the Mellanox error code to common error code before sending to xcvrd

**Comment by keboliu** [comment]
> > ```
> >         return True
> > ```
> 
> return found != 0;
> 
> Refers to: platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py:268 in [3cef25c](https://github.com/Azure/sonic-buildimage/commit/3cef25ca3d0e9c874d70d88d0fce2c694535f29a). [](commit_id = [3cef25c](https://github.com/Azure/sonic-buildimage/commit/3cef25ca3d0e9c874d70d88d0fce2c694535f29a), deletion_comment = False)

revised, thanks.

**Comment by keboliu** [comment]
> > ```
> >             if not port_dict == {}:
> > ```
> 
> If not bool(port_dict):
> 
> Refers to: platform/mellanox/mlnx-platform-api/sonic_platform/chassis.py:451 in [3cef25c](https://github.com/Azure/sonic-buildimage/commit/3cef25ca3d0e9c874d70d88d0fce2c694535f29a). [](commit_id = [3cef25c](https://github.com/Azure/sonic-buildimage/commit/3cef25ca3d0e9c874d70d88d0fce2c694535f29a), deletion_comment = False)

revised, thanks.

File: platform/mellanox/mlnx-platform-api/sonic_platform/chassis.py [hunk 9]
```diff
@@ -448,25 +448,12 @@ def get_change_event(self, timeout=0):
             timeout = MAX_SELECT_DELAY
             while True:
                 status = self.sfp_event.check_sfp_status(port_dict, timeout)
-                if not port_dict == {}:
+                if not bool(port_dict):
                     break
```
Comment (jleveque) [inline platform/mellanox/mlnx-platform-api/sonic_platform/chassis.py:452]: You want to break here if the dictionary is *not* empty, correct? If so, this should read:

```python
                if bool(port_dict):
                    break
```

Comment (keboliu) [reply platform/mellanox/mlnx-platform-api/sonic_platform/chassis.py]: you are absolutely correct.

**Comment by liat-grozovik** [comment]
retest mellanox please

File: dockers/docker-platform-monitor/Dockerfile.j2 [hunk 10]
```diff
@@ -18,7 +18,8 @@ RUN apt-get update &&   \
         rrdtool         \
         python-smbus    \
         ethtool         \
-        dmidecode
+        dmidecode &&    \
+    pip install enum34
```
Comment (jleveque) [inline dockers/docker-platform-monitor/Dockerfile.j2:22]: It doesn't appear this is being used. Am I missing something?

Comment (jleveque) [reply dockers/docker-platform-monitor/Dockerfile.j2:22]: Nevermind. I see this is used in https://github.com/Azure/sonic-platform-daemons/pull/52.

**Pull request merged**